htlcswitch/quiescer.go

```diff
@@ -0,0 +1,233 @@
+func newQuiescer(cfg quiescerCfg) quiescer {
+	return quiescer{
+		cfg: cfg,
+	}
+}
+
+// recvStfu is called when we receive an Stfu message from the remote.
+func (q *quiescer) recvStfu(msg lnwire.Stfu) error {
+	// At the time of this writing, this check that we have already received
+	// an Stfu is not strictly necessary, according to the specification.
+	// However, it is fishy if we do and it is unclear how we should handle
+	// such a case so we will err on the side of caution.
+	if q.received {
+		return fmt.Errorf("%w for channel %v", ErrStfuAlreadyRcvd,
+			q.cfg.chanID)
+	}
+
+	if !q.canRecvStfu() {
+		return fmt.Errorf("%w for channel %v", ErrPendingRemoteUpdates,
+			q.cfg.chanID)
+	}
+
+	q.received = true
+
+	// If the remote party sets the initiator bit to true then we will
+	// remember that they are making a claim to the initiator role. This
+	// does not necessarily mean they will get it, though.
+	q.remoteInit = msg.Initiator
+
+	return nil
+}
+
+// makeStfu is called when we are ready to send an Stfu message. It returns the
+// Stfu message to be sent.
+func (q *quiescer) makeStfu() fn.Result[lnwire.Stfu] {
+	if q.sent {
+		return fn.Errf[lnwire.Stfu]("%w for channel %v",
+			ErrStfuAlreadySent, q.cfg.chanID)
+	}
+
+	if !q.canSendStfu() {
+		return fn.Errf[lnwire.Stfu]("%w for channel %v",
+			ErrPendingLocalUpdates, q.cfg.chanID)
+	}
+
+	stfu := lnwire.Stfu{
+		ChanID:    q.cfg.chanID,
+		Initiator: q.localInit,
+	}
+
+	return fn.Ok(stfu)
+}
+
+// oweStfu returns true if we owe the other party an Stfu. We owe the remote an
+// Stfu when we have received but not yet sent an Stfu, or we are the initiator
+// but have not yet sent an Stfu.
+func (q *quiescer) oweStfu() bool {
+	return q.received && !q.sent
+}
+
+// needStfu returns true if the remote owes us an Stfu. They owe us an Stfu when
+// we have sent but not yet received an Stfu.
+func (q *quiescer) needStfu() bool {
+	return q.sent && !q.received
+}
+
+// isQuiescent returns true if the state machine has been driven all the way to
+// completion. If this returns true, processes that depend on channel quiescence
+// may proceed.
+func (q *quiescer) isQuiescent() bool {
+	return q.sent && q.received
+}
+
+// quiescenceInitiator determines which ChannelParty is the initiator of
+// quiescence for the purposes of downstream protocols. If the channel is not
+// currently quiescent, this method will return ErrNoDownstreamLeader.
+func (q *quiescer) quiescenceInitiator() fn.Result[lntypes.ChannelParty] {
+	switch {
+	case !q.isQuiescent():
+		return fn.Err[lntypes.ChannelParty](ErrNoQuiescenceInitiator)
+
+	case q.localInit && q.remoteInit:
+		// In the case of a tie, the channel initiator wins.
+		return fn.Ok(q.cfg.channelInitiator)
+
+	case !q.localInit && !q.remoteInit:
+		// We assume it is impossible for both to be false if the
+		// channel is quiescent.
+		panic("impossible: one party must have initiated quiescence")
```

Comment:
We should remove all panics, and just return the error case (which also allows one to test this case more easily). I understand the philosophy here, but in practice: we don't want a bug in this area to bring down the entire daemon, causing what's indistinguishable from a DoS (remote peer seems some bug in the logic and can crash `lnd` nodes at will).

Reply:
The issue then is that any switch case that should be impossible by construction has to return errors. This massively bloats the source code (due to `if err != nil { return err }`). Even so, the errors themselves don't actually prevent the node from going down, it just allows a failed assumption to propagate throughout the lifecycle of the program which creates far more insidious issues that may present in ways that are quite distant from the original source. With panics we can halt execution at the source of the issue.

All of that said, I am sympathetic to the idea that LND needs to be robust against failures, to the extent that it actually can, and in this case we wouldn't be compromising the type sig of the function as it already has an error case. However, in the case where we need to enforce the exhaustion of switch cases, I think it is appropriate to dump a panic there since this is technically something the compiler should be able to rule out, but doesn't.

Reply:
> This massively bloats the source code (due to if err != nil { return err }

IMO, that's a minor cost to pay for the assurance that we aren't leaving in place some non-obvious remote crashing DoS vector. 

> However, in the case where we need to enforce the exhaustion of switch cases, I think it is appropriate to dump a panic there since this is technically something the compiler should be able to rule out, but doesn't.

Why can't it just return an error? If unhandled that would potentially just halt a single link, instead of crashing the _entire_ daemon.

Reply:
> Why can't it just return an error? If unhandled that would potentially just halt a single link, instead of crashing the entire daemon.

In this case it's fine but the main issue I'm trying to raise is that if we take this approach everywhere, you will have situations where otherwise pure, total functions now need to return errors because they chose to use a switch statement.